File: backend/models/ticket.py
from pydantic import BaseModel, Field
from typing import Literal
from enum import Enum


class SeverityEnum(str, Enum):
    LOW = "LOW"
    MEDIUM = "MEDIUM"
    HIGH = "HIGH"
    CRITICAL = "CRITICAL"


class UserTypeEnum(str, Enum):
    service_account = "service_account"
    standard_user = "standard_user"
    admin_user = "admin_user"


class TicketIn(BaseModel):
    ticket_id: str
    severity: SeverityEnum
    status: str = "OPEN"
    created_time: str
    rule_triggered: str
    mitre_attack: str
    user: str
    user_type: UserTypeEnum
    source_asset: str
    source_ip: str
    target_asset: str
    target_ip: str
    process: str
    command_line: str
    decoded_command: str
    hour_of_day: int = Field(ge=0, le=23)
    day_of_week: str
    historical_tp_count: int = Field(ge=0)
    historical_fp_count: int = Field(ge=0)


REQUIRED_EXCEL_COLUMNS = [
    "ticket_id",
    "severity",
    "status",
    "created_time",
    "rule_triggered",
    "mitre_attack",
    "user",
    "user_type",
    "source_asset",
    "source_ip",
    "target_asset",
    "target_ip",
    "process",
    "command_line",
    "decoded_command",
    "hour_of_day",
    "day_of_week",
    "historical_tp_count",
    "historical_fp_count",
]

File: frontend/src/types/ticket.ts
export interface Ticket {
  ticket_id: string
  severity: 'LOW' | 'MEDIUM' | 'HIGH' | 'CRITICAL'
  status: string
  created_time: string
  rule_triggered: string
  mitre_attack: string
  user: string
  user_type: 'service_account' | 'standard_user' | 'admin_user'
  source_asset: string
  source_ip: string
  target_asset: string
  target_ip: string
  process: string
  command_line: string
  decoded_command: string
  hour_of_day: number
  day_of_week: string
  historical_tp_count: number
  historical_fp_count: number
}

export interface SimilarIncident {
  ticket_id: string
  similarity: number
  verdict: string
}

export interface GuardrailStatus {
  presidio_pii_scan: string
  nemo_input_rail: string
  nemo_output_rail: string
}

export interface TriageResult {
  ticket_id: string
  verdict: 'TRUE_POSITIVE' | 'FALSE_POSITIVE' | 'NEEDS_REVIEW'
  confidence: number
  xgboost_score: number
  llm_reasoning: string
  root_cause?: string
  contributing_factors?: string[]
  mitre_attack: string
  risk_score: number
  sir_report: string
  processing_time_ms: number
  similar_past_incidents: SimilarIncident[]
  guardrail_status?: GuardrailStatus
}

export interface TriageSummary {
  total: number
  true_positive: number
  false_positive: number
  needs_review: number
}

export interface TriageResponse {
  results: TriageResult[]
  summary: TriageSummary
  errors?: { row: number; ticket_id: string; error: string }[]
}
